File: src/05.linkList/10.2.go
package linklist

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func AddTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var curNode1, curNode2, curNode, newHead *ListNode = l1, l2, nil, nil
	prefix := 0

	for curNode1 != nil || curNode2 != nil {
		num1 := 0
		num2 := 0

		if curNode1 != nil {
			num1 = curNode1.Val
			curNode1 = curNode1.Next
		}

		if curNode2 != nil {
			num2 = curNode2.Val
			curNode2 = curNode2.Next
		}

		total := num1 + num2 + prefix
		value := total % 10
		prefix = total / 10

		newNode := &ListNode{
			Val: value,
		}

		if newHead == nil {
			newHead = newNode
		} else {
			curNode.Next = newNode
		}
		curNode = newNode
	}

	if prefix != 0 {
		curNode.Next = &ListNode{
			Val: prefix,
		}
	}

	return newHead
}

File: src/05.linkList/10.2.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

import { ListNode } from "./01.203";

export function addTwoNumbers(
    l1: ListNode | null,
    l2: ListNode | null
): ListNode | null {
    let curNode1 = l1;
    let curNode2 = l2;
    let prefix = 0;
    let newHead: ListNode | null = null;
    let curNode: ListNode | null = null;
    while (curNode1 || curNode2) {
        let num1 = 0;
        if (curNode1) {
            num1 = curNode1.val;
            curNode1 = curNode1.next;
        }

        let num2 = 0;
        if (curNode2) {
            num2 = curNode2.val;
            curNode2 = curNode2.next;
        }

        let total = num1 + num2 + prefix;
        let value = total % 10;
        prefix = Math.floor(total / 10);

        let newNode = new ListNode(value);
        if (!newHead) {
            newHead = newNode;
        } else {
            curNode.next = newNode;
        }
        curNode = newNode;
    }

    if (prefix) {
        curNode.next = new ListNode(prefix);
    }

    return newHead;
}
